pkg/controller/backupinfrastructure/backup_infrastructure_control.go

```diff
@@ -0,0 +1,372 @@
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return true, err
+	}
+
+	var (
+		backupInfrastructure       = obj.DeepCopy()
+		operationID                = utils.GenerateRandomString(8)
+		backupInfrastructureLogger = logger.NewBackupInfrastructureLogger(logger.Logger, backupInfrastructure.Name, backupInfrastructure.Namespace, operationID)
+		lastOperation              = backupInfrastructure.Status.LastOperation
+		operationType              = computeOperationType(lastOperation)
+	)
+
+	logger.Logger.Infof("[BACKUPINFRASTRUCTURE RECONCILE] %s", key)
+	backupInfrastructureJSON, _ := json.Marshal(backupInfrastructure)
+	backupInfrastructureLogger.Debugf(string(backupInfrastructureJSON))
+
+	op, err := operation.NewWithBackupInfrastructure(backupInfrastructure, backupInfrastructureLogger, c.k8sGardenClient, c.k8sGardenInformers, c.identity, c.secrets, c.imageVector)
+	if err != nil {
+		backupInfrastructureLogger.Errorf("Could not initialize a new operation: %s", err.Error())
+		return true, err
+	}
+
+	// The deletionTimestamp labels a BackupInfrastructure as intended to get deleted. Before deletion,
+	// it has to be ensured that no Infrastructure resources are depending on the BackupInfrastructure anymore.
+	// When this happens the controller will remove the finalizer from the BackupInfrastructure so that it can be garbage collected.
+	if backupInfrastructure.DeletionTimestamp != nil {
+
+		gracePeriod := time.Hour * 24 * time.Duration(*backupInfrastructure.Spec.DeletionGracePeriodDays)
+		if time.Now().Sub(backupInfrastructure.DeletionTimestamp.Time) > gracePeriod {
+			c.recorder.Eventf(backupInfrastructure, corev1.EventTypeNormal, gardenv1beta1.EventDeleting, "[%s] Deleting Backup Infrastructure", operationID)
+			if updateErr := c.updateBackupInfrastructureStatus(op, gardenv1beta1.ShootLastOperationStateProcessing, gardenv1beta1.ShootLastOperationTypeDelete, "Deletion of Backup Infrastructure in progress.", 1, nil); updateErr != nil {
+				backupInfrastructureLogger.Errorf("Could not update the BackupInfrastructure status after deletion start: %+v", updateErr)
+				return true, updateErr
+			}
+
+			if deleteErr := c.deleteBackupInfrastructure(op); deleteErr != nil {
+				c.recorder.Eventf(backupInfrastructure, corev1.EventTypeWarning, gardenv1beta1.EventDeleteError, "[%s] %s", operationID, deleteErr.Description)
+				if updateErr := c.updateBackupInfrastructureStatus(op, gardenv1beta1.ShootLastOperationStateError, gardenv1beta1.ShootLastOperationTypeDelete, deleteErr.Description+" Operation will be retried.", 1, deleteErr); updateErr != nil {
+					backupInfrastructureLogger.Errorf("Could not update the BackupInfrastructure status after deletion error: %+v", updateErr)
+					return true, updateErr
+				}
+				return true, errors.New(deleteErr.Description)
+			}
+			c.recorder.Eventf(backupInfrastructure, corev1.EventTypeNormal, gardenv1beta1.EventDeleted, "[%s] Deleted Backup Infrastructure", operationID)
```

Comment:
Didn't we say that we only want to send events in case of errors?

Reply:
Successful deletion is one time activity and it makes sense to send event in such case.